### PR TITLE
3.0.0

### PR DESCRIPTION
**Pull request by jonathantneal**
- Added: postcss-value-parser v3.0.2
- Added: `method` option to determine how the filter is included
- Added: `filter` option to use the old Internet Explorer proprietary
filter
- Updated: Rewrite of plugin following latest plugin standards
- Updated: Many more properties with rgba values are converted
- Updated: Hexadecimal values are lowercase
- Updated: Tests and documentation

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1 +1,5 @@
+sudo: false
 language: node_js
+node_js:
```
Comment (TrySound) [inline .travis.yml]: Add please stable

**Comment by TrySound** [comment]
Changin test engine is not very good. I like tape more too.

**Comment by TrySound** [comment]
And maybe gulpfile is a little extra for small plugin. I like more npm scripts

File: index.js [hunk 2]
```diff
@@ -1,72 +1,53 @@
...
+				var value = parser(decl.value).walk(function (node) {
+					if (node.type === 'function' && node.value === 'rgba') {
+						var rgba = node.nodes.filter(function (word) {
+							return word.type === 'word';
+						}).map(function (word, index) {
+							return (0 + Math.round(word.value * (index > 2 ? 255 : 1)).toString(16)).slice(-2);
+						});
+
+						node.value = filter && /^background/i.test(decl.prop) ?
+						'progid:DXImageTransform.Microsoft.gradient(GradientType=0,startColorstr=\'#@\',endColorstr=\'#@\')'.replace(/@/g, rgba.slice(-1).concat(rgba.slice(0, -1)).join('')) :
```
Comment (TrySound) [inline index.js]: Better simple string concat, I think.

File: index.js [hunk 3]
```diff
@@ -1,72 +1,53 @@
...
+				return cache[decl.prop] === 1;
+			}).forEach(function (decl) {
+				// transpile rgba values
+				var value = parser(decl.value).walk(function (node) {
+					if (node.type === 'function' && node.value === 'rgba') {
+						var rgba = node.nodes.filter(function (word) {
+							return word.type === 'word';
+						}).map(function (word, index) {
+							return (0 + Math.round(word.value * (index > 2 ? 255 : 1)).toString(16)).slice(-2);
+						});
```
Comment (TrySound) [inline index.js]: Double iterating is not fast.
```js
node.nodes[0].value,
node.nodes[2].value,
node.nodes[4].value,
```
This much faster

File: .eslintrc [hunk 4]
```diff
@@ -1,35 +1,122 @@
----
```
Comment (MoOx) [inline .eslintrc]: why do you change this ?

File: .gitignore [hunk 5]
```diff
@@ -1,4 +1,2 @@
-test/fixtures/*.actual.css
```
Comment (MoOx) [inline .gitignore]: why?

File: .travis.yml (hunk 1 above)
Comment (MoOx) [inline .travis.yml]: +1

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
+You want to help? You rock! Now, take a moment to be sure your contributions make sense to everyone else.
```
Comment (MoOx) [inline CONTRIBUTING.md]: For simple project, this kind of file are not really required...

File: gulpfile.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+	return gulp.src('test/*.js', { read: false })
+		.pipe(mocha()).on('error', done);
+});
+
+gulp.task('default', ['lint', 'test']);
+
+gulp.task('watch', function() {
+	gulp.watch(files, ['lint', 'test']);
+});
```
Comment (MoOx) [inline gulpfile.js]: gulp is just overkill here...

File: test/test.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		expect(actualCSS).to.eql(expectCSS);
+		expect(result.warnings()).to.have.length(warnings);
+
+		done();
+	}).catch(function (error) {
+		done(error);
+	});
+}
+
+describe('postcss-color-rgba-fallback', function () {
```
Comment (MoOx) [inline test/test.js]: lol why did you change tape to mocha ?
Tip: https://medium.com/javascript-scene/why-i-use-tape-instead-of-mocha-so-should-you-6aa105d8eaf4

Comment (jonathantneal) [inline test/test.js]: This is just a fixtured version of https://github.com/postcss/postcss-plugin-boilerplate/blob/master/test/test.js

File: .eslintrc (hunk 4 above)
Comment (jonathantneal) [inline .eslintrc]: https://github.com/postcss/postcss-plugin-boilerplate/blob/master/.eslintrc

Because it’s a PostCSS plugin and there is a PostCSS plugin boilerplate .eslintrc

File: .gitignore (hunk 5 above)
Comment (jonathantneal) [inline .gitignore]: Again, just following https://github.com/postcss/postcss-plugin-boilerplate/blob/master/.gitignore

File: CONTRIBUTING.md (hunk 6 above)
Comment (jonathantneal) [inline CONTRIBUTING.md]: Any project with an issue tracker deserves it, especially one that has had activity. It shows up automatically in the issue tracker and many developers actually find it very helpful.

File: index.js (hunk 2 above)
Comment (jonathantneal) [inline index.js]: That’s fair. I will change this.

File: index.js (hunk 3 above)
Comment (jonathantneal) [inline index.js]: It’s not fast, but it’s more *correct*. Still, I suppose we can make this presumption so I will change this.

File: .travis.yml (hunk 1 above)
Comment (jonathantneal) [inline .travis.yml]: Yes, I will add this.

File: test/test.js (hunk 8 above)
Comment (TrySound) [inline test/test.js]: Boilerplate is starter kit, which can be much different from end of project. It's not in guidelines, so developer can choose what he want.
We want tape :))

Comment (MoOx) [inline test/test.js]: +1, boilerplate is just a starting point for newbies.

Comment (jonathantneal) [inline test/test.js]: :disappointed:

File: index.js [hunk 9]
```diff
@@ -1,72 +1,53 @@
...
+
+			rule.nodes.filter(function (decl) {
+				// assure property is cached
+				cache[decl.prop] = cache[decl.prop] || 0;
+
+				// count each time property is cached
+				++cache[decl.prop];
+
+				// keep valid, non-repeated, rgba-containing properties
+				return cache[decl.prop] === 1 && props.indexOf(decl.prop) !== -1 && /rgba/i.test(decl.value);
```
Comment (ben-eb) [inline index.js]: These are nicer alternatives to the second condition. :smile:

```js
props.indexOf(decl.prop) > -1
```

or:

```js
~props.indexOf(decl.prop)
```

File: test/test.js (hunk 8 above)
Comment (magsout) [inline test/test.js]: Agree with @MoOx and @TrySound

**Comment by magsout** [comment]
Thanks but I PR of @TrySound is better. Plus, there are too many changes from the original code
